coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+### [Mutable References](#mutable-references)
+Use `let` instead of `var` if you need a mutable reference.
+> `let` is block-scoped rather than function scoped
+
+# Objects
+### [Object Instansiation](#object-instansiation)
+
+#### Literal Syntax
+Literal syntax is prefered when dealing with simple cases. Perfectly valid usage of literal syntax:
+```typescript
+interface User {
+  username: string;
+  age: number;
+}
+
+const user: User = {
+  username: 'kanedoe',
+  age: 42
+}
+```
+**Avoid!!!** using literal syntax if function are defined in the object!
+```typescript
+interface Store {
+  name: string;
+  isOpen(hour24: number): boolean;
+}
+
+const isOpen = (hour24: number) => {
+  return hour24 > 6 && hour24 < 20;
+}
+
+const butcherShop = {
+  name: 'Meaty',
+  isOpen // AVOID!
+}
+```
+
+#### Interfaces
+Interfaces can be created through literal syntax as long as the following rules apply.
+- The interface does not contain any functions.
+- The interface is a mock-up used for a test class.
+
+If an interface contains a function you should instead make a class implementing said interface for instantiation. 
+
+#### Classes
+Instead of using literal syntax on complex interfaces, create a class which implements the interface.
+```typescript
+interface Store {
+  readonly name: string;
+  isOpen(hour24: number): boolean;
+}
+
+class ButcherShop implements Store {
+  private readonly _name: string;
+  private readonly _openingHour: number;
+  private readonly _closingHour: number;
+
+  constructor(name: string, openingHour: number, closingHour: number) {
+    this._name = name;
+    this._openingHour = openingHour;
+    this._closingHour = closingHour;
+  }
+
+  get name(): string {
+    return this._name;
+  }
+
+  isOpen(hour24: number): boolean {
+    return hour24 >= this._openingHour && hour24 < this._closingHour;
+  }
+}
+```
+> **Note:** Make sure to create the correct access level of variables. The Store in the example above has a readonly name to prevent the name of the shop to be changed, in this case where the readonly variable is a primitive it gets passed by value.
+
+> **Note:** Place readonly variables above mutable variables in classes.
+### [Getters and Setters](#getters-and-setters)
+Getters and setters are used to make sure that data is not manipulated in the wrong way. In the example above a getter on the name is safe since the variable itself is passed by reference. But what happens when the getter returns an object?
+```typescript
+interface Owner {
+  name: string;
+}
+
+interface Store {
+  readonly name: string;
+  readonly owner: Owner;
+}
+
+class CandyShop implements Store {
+  private readonly _name: string;
+  private readonly _owner: Owner;
+
+  constructor(name: string, owner: Owner) {
+    this._name = name;
+    this._owner = owner;
+  }
+
+  get name(): string {
+    return this._name;
+  }
+
+  get owner(): Owner {
+    return this._owner;
+  }
```

Comment:
I think this is perfectly valid. This is how e.g. C# works too and should be familiar. If we want to be explicit about this sort of stuff, we should return `[Readonly<Owner>](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlyt)` instead.